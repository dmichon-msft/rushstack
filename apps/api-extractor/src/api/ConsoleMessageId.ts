// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

/**
 * Unique identifiers for console messages reported by API Extractor.
 *
 * @remarks
 *
 * These strings are possible values for the {@link ExtractorMessage.messageId} property
 * when the `ExtractorMessage.category` is {@link ExtractorMessageCategory.Console}.
 *
 * @public
 */
// eslint-disable-next-line @typescript-eslint/typedef
export const ConsoleMessageId = {
  /**
   * "Analysis will use the bundled TypeScript version ___"
   */
  Preamble: 'console-preamble',

  /**
   * "The target project appears to use TypeScript ___ which is newer than the bundled compiler engine;
   * consider upgrading API Extractor."
   */
  CompilerVersionNotice: 'console-compiler-version-notice',

  /**
   * "Using custom TSDoc config from ___"
   */
  UsingCustomTSDocConfig: 'console-using-custom-tsdoc-config',

  /**
   * "Found metadata in ___"
   */
  FoundTSDocMetadata: 'console-found-tsdoc-metadata',

  /**
   * "Writing: ___"
   */
  WritingDocModelFile: 'console-writing-doc-model-file',

  /**
   * "Writing package typings: ___"
   */
  WritingDtsRollup: 'console-writing-dts-rollup',

  /**
   * "You have changed the public API signature for this project.
   * Please copy the file ___ to ___, or perform a local build (which does this automatically).
   * See the Git repo documentation for more info."
   *
   * OR
   *
   * "The API report file is missing.
   * Please copy the file ___ to ___, or perform a local build (which does this automatically).
   * See the Git repo documentation for more info."
   */
  ApiReportNotCopied: 'console-api-report-not-copied',

  /**
   * "You have changed the public API signature for this project.  Updating ___"
   */
  ApiReportCopied: 'console-api-report-copied',

  /**
   * "The API report is up to date: ___"
   */
  ApiReportUnchanged: 'console-api-report-unchanged',

  /**
   * "The API report file was missing, so a new file was created. Please add this file to Git: ___"
   */
  ApiReportCreated: 'console-api-report-created',

  /**
   * "Unable to create the API report file. Please make sure the target folder exists: ___"
   */
  ApiReportFolderMissing: 'console-api-report-folder-missing',

  /**
   * Used for the information printed when the "--diagnostics" flag is enabled.
   */
  Diagnostics: 'console-diagnostics'
} as const;
// eslint-disable-next-line @typescript-eslint/no-namespace
export declare namespace ConsoleMessageId {
  /**
   * "Analysis will use the bundled TypeScript version ___"
   */
  export type Preamble = typeof ConsoleMessageId.Preamble;

  /**
   * "The target project appears to use TypeScript ___ which is newer than the bundled compiler engine;
   * consider upgrading API Extractor."
   */
  export type CompilerVersionNotice = typeof ConsoleMessageId.CompilerVersionNotice;

  /**
   * "Using custom TSDoc config from ___"
   */
  export type UsingCustomTSDocConfig = typeof ConsoleMessageId.UsingCustomTSDocConfig;

  /**
   * "Found metadata in ___"
   */
  export type FoundTSDocMetadata = typeof ConsoleMessageId.FoundTSDocMetadata;

  /**
   * "Writing: ___"
   */
  export type WritingDocModelFile = typeof ConsoleMessageId.WritingDocModelFile;

  /**
   * "Writing package typings: ___"
   */
  export type WritingDtsRollup = typeof ConsoleMessageId.WritingDtsRollup;

  /**
   * "You have changed the public API signature for this project.
   * Please copy the file ___ to ___, or perform a local build (which does this automatically).
   * See the Git repo documentation for more info."
   *
   * OR
   *
   * "The API report file is missing.
   * Please copy the file ___ to ___, or perform a local build (which does this automatically).
   * See the Git repo documentation for more info."
   */
  export type ApiReportNotCopied = typeof ConsoleMessageId.ApiReportNotCopied;

  /**
   * "You have changed the public API signature for this project.  Updating ___"
   */
  export type ApiReportCopied = typeof ConsoleMessageId.ApiReportCopied;

  /**
   * "The API report is up to date: ___"
   */
  export type ApiReportUnchanged = typeof ConsoleMessageId.ApiReportUnchanged;

  /**
   * "The API report file was missing, so a new file was created. Please add this file to Git: ___"
   */
  export type ApiReportCreated = typeof ConsoleMessageId.ApiReportCreated;

  /**
   * "Unable to create the API report file. Please make sure the target folder exists: ___"
   */
  export type ApiReportFolderMissing = typeof ConsoleMessageId.ApiReportFolderMissing;

  /**
   * Used for the information printed when the "--diagnostics" flag is enabled.
   */
  export type Diagnostics = typeof ConsoleMessageId.Diagnostics;
}
export type ConsoleMessageId = typeof ConsoleMessageId[keyof typeof ConsoleMessageId];
