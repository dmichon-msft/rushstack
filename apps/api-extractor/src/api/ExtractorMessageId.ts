// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

/**
 * Unique identifiers for messages reported by API Extractor during its analysis.
 *
 * @remarks
 *
 * These strings are possible values for the {@link ExtractorMessage.messageId} property
 * when the `ExtractorMessage.category` is {@link ExtractorMessageCategory.Extractor}.
 *
 * @public
 */
// eslint-disable-next-line @typescript-eslint/typedef
export const ExtractorMessageId = {
  /**
   * "The doc comment should not contain more than one release tag."
   */
  ExtraReleaseTag: 'ae-extra-release-tag',

  /**
   * "This symbol has another declaration with a different release tag."
   */
  DifferentReleaseTags: 'ae-different-release-tags',

  /**
   * "The symbol ___ is marked as ___, but its signature references ___ which is marked as ___."
   */
  IncompatibleReleaseTags: 'ae-incompatible-release-tags',

  /**
   * "___ is exported by the package, but it is missing a release tag (`@alpha`, `@beta`, `@public`, or `@internal`)."
   */
  MissingReleaseTag: 'ae-missing-release-tag',

  /**
   * "The `@packageDocumentation` comment must appear at the top of entry point *.d.ts file."
   */
  MisplacedPackageTag: 'ae-misplaced-package-tag',

  /**
   * "The symbol ___ needs to be exported by the entry point ___."
   */
  ForgottenExport: 'ae-forgotten-export',

  /**
   * "The name ___ should be prefixed with an underscore because the declaration is marked as `@internal`."
   */
  InternalMissingUnderscore: 'ae-internal-missing-underscore',

  /**
   * "Mixed release tags are not allowed for ___ because one of its declarations is marked as `@internal`."
   */
  InternalMixedReleaseTag: 'ae-internal-mixed-release-tag',

  /**
   * "The `@preapproved` tag cannot be applied to ___ because it is not a supported declaration type."
   */
  PreapprovedUnsupportedType: 'ae-preapproved-unsupported-type',

  /**
   * "The `@preapproved` tag cannot be applied to ___ without an `@internal` release tag."
   */
  PreapprovedBadReleaseTag: 'ae-preapproved-bad-release-tag',

  /**
   * "The `@inheritDoc` reference could not be resolved."
   */
  UnresolvedInheritDocReference: 'ae-unresolved-inheritdoc-reference',

  /**
   * "The `@inheritDoc` tag needs a TSDoc declaration reference; signature matching is not supported yet."
   *
   * @privateRemarks
   * In the future, we will implement signature matching so that you can write `{@inheritDoc}` and API Extractor
   * will find a corresponding member from a base class (or implemented interface).  Until then, the tag
   * always needs an explicit declaration reference such as `{@inhertDoc MyBaseClass.sameMethod}`.
   */
  UnresolvedInheritDocBase: 'ae-unresolved-inheritdoc-base',

  /**
   * "The `@inheritDoc` tag for ___ refers to its own declaration."
   */
  CyclicInheritDoc: 'ae-cyclic-inherit-doc',

  /**
   * "The `@link` reference could not be resolved."
   */
  UnresolvedLink: 'ae-unresolved-link',

  /**
   * "The doc comment for the property ___ must appear on the getter, not the setter."
   */
  SetterWithDocs: 'ae-setter-with-docs',

  /**
   * "The property ___ has a setter but no getter."
   */
  MissingGetter: 'ae-missing-getter'
} as const;
// eslint-disable-next-line @typescript-eslint/no-namespace
export declare namespace ExtractorMessageId {
  /**
   * "The doc comment should not contain more than one release tag."
   */
  export type ExtraReleaseTag = typeof ExtractorMessageId.ExtraReleaseTag;

  /**
   * "This symbol has another declaration with a different release tag."
   */
  export type DifferentReleaseTags = typeof ExtractorMessageId.DifferentReleaseTags;

  /**
   * "The symbol ___ is marked as ___, but its signature references ___ which is marked as ___."
   */
  export type IncompatibleReleaseTags = typeof ExtractorMessageId.IncompatibleReleaseTags;

  /**
   * "___ is exported by the package, but it is missing a release tag (`@alpha`, `@beta`, `@public`, or `@internal`)."
   */
  export type MissingReleaseTag = typeof ExtractorMessageId.MissingReleaseTag;

  /**
   * "The `@packageDocumentation` comment must appear at the top of entry point *.d.ts file."
   */
  export type MisplacedPackageTag = typeof ExtractorMessageId.MisplacedPackageTag;

  /**
   * "The symbol ___ needs to be exported by the entry point ___."
   */
  export type ForgottenExport = typeof ExtractorMessageId.ForgottenExport;

  /**
   * "The name ___ should be prefixed with an underscore because the declaration is marked as `@internal`."
   */
  export type InternalMissingUnderscore = typeof ExtractorMessageId.InternalMissingUnderscore;

  /**
   * "Mixed release tags are not allowed for ___ because one of its declarations is marked as `@internal`."
   */
  export type InternalMixedReleaseTag = typeof ExtractorMessageId.InternalMixedReleaseTag;

  /**
   * "The `@preapproved` tag cannot be applied to ___ because it is not a supported declaration type."
   */
  export type PreapprovedUnsupportedType = typeof ExtractorMessageId.PreapprovedUnsupportedType;

  /**
   * "The `@preapproved` tag cannot be applied to ___ without an `@internal` release tag."
   */
  export type PreapprovedBadReleaseTag = typeof ExtractorMessageId.PreapprovedBadReleaseTag;

  /**
   * "The `@inheritDoc` reference could not be resolved."
   */
  export type UnresolvedInheritDocReference = typeof ExtractorMessageId.UnresolvedInheritDocReference;

  /**
   * "The `@inheritDoc` tag needs a TSDoc declaration reference; signature matching is not supported yet."
   *
   * @privateRemarks
   * In the future, we will implement signature matching so that you can write `{@inheritDoc}` and API Extractor
   * will find a corresponding member from a base class (or implemented interface).  Until then, the tag
   * always needs an explicit declaration reference such as `{@inhertDoc MyBaseClass.sameMethod}`.
   */
  export type UnresolvedInheritDocBase = typeof ExtractorMessageId.UnresolvedInheritDocBase;

  /**
   * "The `@inheritDoc` tag for ___ refers to its own declaration."
   */
  export type CyclicInheritDoc = typeof ExtractorMessageId.CyclicInheritDoc;

  /**
   * "The `@link` reference could not be resolved."
   */
  export type UnresolvedLink = typeof ExtractorMessageId.UnresolvedLink;

  /**
   * "The doc comment for the property ___ must appear on the getter, not the setter."
   */
  export type SetterWithDocs = typeof ExtractorMessageId.SetterWithDocs;

  /**
   * "The property ___ has a setter but no getter."
   */
  export type MissingGetter = typeof ExtractorMessageId.MissingGetter;
}
export type ExtractorMessageId = typeof ExtractorMessageId[keyof typeof ExtractorMessageId];

export const allExtractorMessageIds: Set<string> = new Set<string>(Object.values(ExtractorMessageId));
