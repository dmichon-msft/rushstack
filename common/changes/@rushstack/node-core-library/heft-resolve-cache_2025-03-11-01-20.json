{
  "changes": [
    {
      "packageName": "@rushstack/node-core-library",
      "comment": "Add `useNodeJSResolver` option to `Import.resolvePackage` to rely on the built-in `require.resolve` and share its cache.",
      "type": "minor"
    }
  ],
  "packageName": "@rushstack/node-core-library"
}