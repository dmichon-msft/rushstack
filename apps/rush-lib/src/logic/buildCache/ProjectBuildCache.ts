// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import * as path from 'path';
import * as events from 'events';
import * as crypto from 'crypto';
import type * as stream from 'stream';
import * as tar from 'tar';
import { FileSystem, Path, Terminal } from '@rushstack/node-core-library';
import * as fs from 'fs';

import { RushConfigurationProject } from '../../api/RushConfigurationProject';
import { ProjectChangeAnalyzer } from '../ProjectChangeAnalyzer';
import { RushProjectConfiguration } from '../../api/RushProjectConfiguration';
import { RushConstants } from '../RushConstants';
import { BuildCacheConfiguration } from '../../api/BuildCacheConfiguration';
import { CloudBuildCacheProviderBase } from './CloudBuildCacheProviderBase';
import { FileSystemBuildCacheProvider } from './FileSystemBuildCacheProvider';
import { TarExecutable } from '../../utilities/TarExecutable';
import { Utilities } from '../../utilities/Utilities';

interface IProjectBuildCacheOptions {
  buildCacheConfiguration: BuildCacheConfiguration;
  projectConfiguration: RushProjectConfiguration;
  command: string;
  trackedProjectFiles: string[] | undefined;
  projectChangeAnalyzer: ProjectChangeAnalyzer;
  terminal: Terminal;
}

export class ProjectBuildCache {
  /**
   * null === we haven't tried to initialize yet
   * undefined === unable to initialize
   */
  private static _tarUtility: TarExecutable | null | undefined = null;

  private readonly _project: RushConfigurationProject;
  private readonly _localBuildCacheProvider: FileSystemBuildCacheProvider;
  private readonly _cloudBuildCacheProvider: CloudBuildCacheProviderBase | undefined;
  private readonly _buildCacheEnabled: boolean;
  private readonly _projectOutputFolderNames: string[];
  private _cacheId: string | undefined;

  private constructor(cacheId: string | undefined, options: IProjectBuildCacheOptions) {
    this._project = options.projectConfiguration.project;
    this._localBuildCacheProvider = options.buildCacheConfiguration.localCacheProvider;
    this._cloudBuildCacheProvider = options.buildCacheConfiguration.cloudCacheProvider;
    this._buildCacheEnabled = options.buildCacheConfiguration.buildCacheEnabled;
    this._projectOutputFolderNames = options.projectConfiguration.projectOutputFolderNames || [];
    this._cacheId = cacheId;
  }

  private static _tryGetTarUtility(terminal: Terminal): TarExecutable | undefined {
    if (ProjectBuildCache._tarUtility === null) {
      ProjectBuildCache._tarUtility = TarExecutable.tryInitialize(terminal);
    }

    return ProjectBuildCache._tarUtility;
  }

  public static async tryGetProjectBuildCache(
    options: IProjectBuildCacheOptions
  ): Promise<ProjectBuildCache | undefined> {
    const { terminal, projectConfiguration, trackedProjectFiles } = options;
    if (!trackedProjectFiles) {
      return undefined;
    }

    if (!ProjectBuildCache._validateProject(terminal, projectConfiguration, trackedProjectFiles)) {
      return undefined;
    }

    const cacheId: string | undefined = await ProjectBuildCache._getCacheId(options);
    return new ProjectBuildCache(cacheId, options);
  }

  private static _validateProject(
    terminal: Terminal,
    projectConfiguration: RushProjectConfiguration,
    trackedProjectFiles: string[]
  ): boolean {
    const normalizedProjectRelativeFolder: string = Path.convertToSlashes(
      projectConfiguration.project.projectRelativeFolder
    );
    const outputFolders: string[] = [];
    if (projectConfiguration.projectOutputFolderNames) {
      for (const outputFolderName of projectConfiguration.projectOutputFolderNames) {
        outputFolders.push(`${normalizedProjectRelativeFolder}/${outputFolderName}/`);
      }
    }

    const inputOutputFiles: string[] = [];
    for (const file of trackedProjectFiles) {
      for (const outputFolder of outputFolders) {
        if (file.startsWith(outputFolder)) {
          inputOutputFiles.push(file);
        }
      }
    }

    if (inputOutputFiles.length > 0) {
      terminal.writeWarningLine(
        'Unable to use build cache. The following files are used to calculate project state ' +
          `and are considered project output: ${inputOutputFiles.join(', ')}`
      );
      return false;
    } else {
      return true;
    }
  }

  public async tryRestoreFromCacheAsync(terminal: Terminal): Promise<boolean> {
    const cacheId: string | undefined = this._cacheId;
    if (!cacheId) {
      terminal.writeWarningLine('Unable to get cache ID. Ensure Git is installed.');
      return false;
    }

    if (!this._buildCacheEnabled) {
      // Skip reading local and cloud build caches, without any noise
      return false;
    }

    let localCacheEntryPath: string | undefined =
      await this._localBuildCacheProvider.tryGetCacheEntryPathByIdAsync(terminal, cacheId);
    let cacheEntryBuffer: Buffer | undefined;
    let updateLocalCacheSuccess: boolean | undefined;
    if (!localCacheEntryPath && this._cloudBuildCacheProvider) {
      terminal.writeVerboseLine(
        'This project was not found in the local build cache. Querying the cloud build cache.'
      );

      cacheEntryBuffer = await this._cloudBuildCacheProvider.tryGetCacheEntryBufferByIdAsync(
        terminal,
        cacheId
      );
      if (cacheEntryBuffer) {
        try {
          localCacheEntryPath = await this._localBuildCacheProvider.trySetCacheEntryBufferAsync(
            terminal,
            cacheId,
            cacheEntryBuffer
          );
          updateLocalCacheSuccess = true;
        } catch (e) {
          updateLocalCacheSuccess = false;
        }
      }
    }

    if (!localCacheEntryPath && !cacheEntryBuffer) {
      terminal.writeVerboseLine('This project was not found in the build cache.');
      return false;
    }

    terminal.writeLine('Build cache hit.');

    const projectFolderPath: string = this._project.projectFolder;

    // Purge output folders
    terminal.writeVerboseLine(`Clearing cached folders: ${this._projectOutputFolderNames.join(', ')}`);
    await Promise.all(
      this._projectOutputFolderNames.map((outputFolderName: string) =>
        FileSystem.deleteFolderAsync(`${projectFolderPath}/${outputFolderName}`)
      )
    );

    const tarUtility: TarExecutable | undefined = ProjectBuildCache._tryGetTarUtility(terminal);
    let restoreSuccess: boolean = false;
    if (tarUtility && localCacheEntryPath) {
      const logFilePath: string = this._getTarLogFilePath();
      const tarExitCode: number = await tarUtility.tryUntarAsync({
        archivePath: localCacheEntryPath,
        outputFolderPath: projectFolderPath,
        logFilePath
      });
      if (tarExitCode === 0) {
        restoreSuccess = true;
      } else {
        terminal.writeWarningLine(
          `"tar" exited with code ${tarExitCode} while attempting to restore cache entry. ` +
            'Rush will attempt to extract from the cache entry with a JavaScript implementation of tar. ' +
            `See "${logFilePath}" for logs from the tar process.`
        );
      }
    }

    if (!restoreSuccess) {
      if (!cacheEntryBuffer && localCacheEntryPath) {
        cacheEntryBuffer = await FileSystem.readFileToBufferAsync(localCacheEntryPath);
      }

      if (!cacheEntryBuffer) {
        throw new Error('Expected the cache entry buffer to be set.');
      }

      // If we don't have tar on the PATH, if we failed to update the local cache entry,
      // or if the tar binary failed, untar in-memory
      const tarStream: stream.Writable = tar.extract({
        cwd: projectFolderPath,
        // Set to true to omit writing mtime value for extracted entries.
        m: true
      });
      try {
        const tarPromise: Promise<unknown> = events.once(tarStream, 'drain');
        tarStream.write(cacheEntryBuffer);
        await tarPromise;
        restoreSuccess = true;
      } catch (e) {
        restoreSuccess = false;
      }
    }

    if (restoreSuccess) {
      terminal.writeLine('Successfully restored output from the build cache.');
    } else {
      terminal.writeWarningLine('Unable to restore output from the build cache.');
    }

    if (updateLocalCacheSuccess === false) {
      terminal.writeWarningLine('Unable to update the local build cache with data from the cloud cache.');
    }

    return restoreSuccess;
  }

  public async trySetCacheEntryAsync(terminal: Terminal): Promise<boolean> {
    const cacheId: string | undefined = this._cacheId;
    if (!cacheId) {
      terminal.writeWarningLine('Unable to get cache ID. Ensure Git is installed.');
      return false;
    }

    if (!this._buildCacheEnabled) {
      // Skip writing local and cloud build caches, without any noise
      return false;
    }

    const projectFolderPath: string = this._project.projectFolder;
    const foldersToCache: string[] | undefined = await this._tryGetFoldersToCacheAsync(terminal);
    if (!foldersToCache) {
      return false;
    }

    terminal.writeVerboseLine(`Caching build output folders: ${foldersToCache.join(', ')}`);

    let localCacheEntryPath: string | undefined;

    const tarUtility: TarExecutable | undefined = ProjectBuildCache._tryGetTarUtility(terminal);
    if (tarUtility) {
      const filesToCache: string[] | undefined = await this._tryCollectFilesToCacheAsync(
        foldersToCache,
        terminal
      );
      if (!filesToCache) {
        return false;
      }
      const tempLocalCacheEntryPath: string = this._localBuildCacheProvider.getCacheEntryPath(cacheId);
      const logFilePath: string = this._getTarLogFilePath();
      const tarExitCode: number = await tarUtility.tryCreateArchiveFromProjectPathsAsync({
        archivePath: tempLocalCacheEntryPath,
        paths: filesToCache,
        project: this._project,
        logFilePath
      });

      if (tarExitCode === 0) {
        localCacheEntryPath = tempLocalCacheEntryPath;
      } else {
        terminal.writeWarningLine(
          `"tar" exited with code ${tarExitCode} while attempting to create the cache entry. ` +
            'Rush will attempt to create the cache entry with a JavaScript implementation of tar. ' +
            `See "${logFilePath}" for logs from the tar process.`
        );
      }
    }

    let cacheEntryBuffer: Buffer | undefined;
    let setLocalCacheEntryPromise: Promise<string> | undefined;
    if (!localCacheEntryPath) {
      const filesToCache: string[] | undefined = await this._tryCollectFilesToCacheAsync(
        foldersToCache,
        terminal
      );
      if (!filesToCache) {
        return false;
      }
      // If we weren't able to create the cache entry with tar, try to do it with the "tar" NPM package
      const tarStream: stream.Readable = tar.create(
        {
          gzip: true,
          portable: true,
          strict: true,
          cwd: projectFolderPath
        },
        filesToCache
      );
      cacheEntryBuffer = await Utilities.readStreamToBufferAsync(tarStream);
      setLocalCacheEntryPromise = this._localBuildCacheProvider.trySetCacheEntryBufferAsync(
        terminal,
        cacheId,
        cacheEntryBuffer
      );
    } else {
      setLocalCacheEntryPromise = Promise.resolve(localCacheEntryPath);
    }

    let setCloudCacheEntryPromise: Promise<boolean> | undefined;

    // Note that "writeAllowed" settings (whether in config or environment) always apply to
    // the configured CLOUD cache. If the cache is enabled, rush is always allowed to read from and
    // write to the local build cache.

    if (this._cloudBuildCacheProvider?.isCacheWriteAllowed) {
      if (!cacheEntryBuffer) {
        if (localCacheEntryPath) {
          cacheEntryBuffer = await FileSystem.readFileToBufferAsync(localCacheEntryPath);
        } else {
          throw new Error('Expected the local cache entry path to be set.');
        }
      }

      setCloudCacheEntryPromise = this._cloudBuildCacheProvider?.trySetCacheEntryBufferAsync(
        terminal,
        cacheId,
        cacheEntryBuffer
      );
    }

    let localCachePath: string;
    let updateCloudCacheSuccess: boolean;
    if (setCloudCacheEntryPromise) {
      [updateCloudCacheSuccess, localCachePath] = await Promise.all([
        setCloudCacheEntryPromise,
        setLocalCacheEntryPromise
      ]);
    } else {
      updateCloudCacheSuccess = true;
      localCachePath = await setLocalCacheEntryPromise;
    }

    const success: boolean = updateCloudCacheSuccess && !!localCachePath;
    if (success) {
      terminal.writeLine('Successfully set cache entry.');
    } else if (!localCachePath && updateCloudCacheSuccess) {
      terminal.writeWarningLine('Unable to set local cache entry.');
    } else if (localCachePath && !updateCloudCacheSuccess) {
      terminal.writeWarningLine('Unable to set cloud cache entry.');
    } else {
      terminal.writeWarningLine('Unable to set both cloud and local cache entries.');
    }

    return success;
  }

  private async _tryGetFoldersToCacheAsync(terminal: Terminal): Promise<string[] | undefined> {
    const filteredOutputFolderNames: string[] = [];

    const { projectFolder } = this._project;
    const contents: fs.Dirent[] = await fs.promises.readdir(projectFolder, {
      encoding: 'utf-8',
      withFileTypes: true
    });

    let hasError: boolean = false;
    const outputFolderPaths: Set<string> = new Set(this._projectOutputFolderNames);
    for (const item of contents) {
      if (outputFolderPaths.has(item.name)) {
        if (item.isDirectory()) {
          filteredOutputFolderNames.push(item.name);
        } else {
          hasError = true;
          terminal.writeError(
            `Unable to include folder "${projectFolder}/${item.name}" in build cache. It is a ${
              item.isSymbolicLink() ? 'symbolic link' : 'file'
            }.`
          );
        }
      }
    }

    return hasError ? undefined : filteredOutputFolderNames;
  }

  private async _tryCollectFilesToCacheAsync(
    filteredOutputFolderNames: string[],
    terminal: Terminal
  ): Promise<string[] | undefined> {
    const projectFolderPath: string = this._project.projectFolder;

    const outputFilePaths: string[] = [];
    const queue: Map<string, string> = new Map();
    for (const outputFolder of filteredOutputFolderNames) {
      queue.set(outputFolder, `${projectFolderPath}/${outputFolder}`);
    }

    for (const [relativePath, diskPath] of queue) {
      const children: fs.Dirent[] = await fs.promises.readdir(diskPath, {
        encoding: 'utf-8',
        withFileTypes: true
      });

      for (const child of children) {
        const childRelativePath: string = `${relativePath}/${child.name}`;
        if (child.isSymbolicLink()) {
          terminal.writeError(
            `Unable to include "${childRelativePath}" in build cache. It is a symbolic link.`
          );
          return undefined;
        } else if (child.isDirectory()) {
          queue.set(childRelativePath, `${diskPath}/${child.name}`);
        } else {
          outputFilePaths.push(childRelativePath);
        }
      }
    }

    return outputFilePaths;
  }

  private _getTarLogFilePath(): string {
    return path.join(this._project.projectRushTempFolder, 'build-cache-tar.log');
  }

  private static async _getCacheId(options: IProjectBuildCacheOptions): Promise<string | undefined> {
    // The project state hash is calculated in the following method:
    // - The current project's hash (see ProjectChangeAnalyzer.getProjectStateHash) is
    //   calculated and appended to an array
    // - The current project's recursive dependency projects' hashes are calculated
    //   and appended to the array
    // - A SHA1 hash is created and the following data is fed into it, in order:
    //   1. The JSON-serialized list of output folder names for this
    //      project (see ProjectBuildCache._projectOutputFolderNames)
    //   2. The command that will be run in the project
    //   3. Each dependency project hash (from the array constructed in previous steps),
    //      in sorted alphanumerical-sorted order
    // - A hex digest of the hash is returned
    const projectChangeAnalyzer: ProjectChangeAnalyzer = options.projectChangeAnalyzer;
    const projectStates: string[] = [];
    const projectsThatHaveBeenProcessed: Set<RushConfigurationProject> = new Set<RushConfigurationProject>();
    let projectsToProcess: Set<RushConfigurationProject> = new Set<RushConfigurationProject>();
    projectsToProcess.add(options.projectConfiguration.project);

    while (projectsToProcess.size > 0) {
      const newProjectsToProcess: Set<RushConfigurationProject> = new Set<RushConfigurationProject>();
      for (const projectToProcess of projectsToProcess) {
        projectsThatHaveBeenProcessed.add(projectToProcess);

        const projectState: string | undefined = await projectChangeAnalyzer._tryGetProjectStateHashAsync(
          projectToProcess.packageName,
          options.terminal
        );
        if (!projectState) {
          // If we hit any projects with unknown state, return unknown cache ID
          return undefined;
        } else {
          projectStates.push(projectState);
          for (const dependency of projectToProcess.dependencyProjects) {
            if (!projectsThatHaveBeenProcessed.has(dependency)) {
              newProjectsToProcess.add(dependency);
            }
          }
        }
      }

      projectsToProcess = newProjectsToProcess;
    }

    const sortedProjectStates: string[] = projectStates.sort();
    const hash: crypto.Hash = crypto.createHash('sha1');
    const serializedOutputFolders: string = JSON.stringify(
      options.projectConfiguration.projectOutputFolderNames
    );
    hash.update(serializedOutputFolders);
    hash.update(RushConstants.hashDelimiter);
    hash.update(options.command);
    hash.update(RushConstants.hashDelimiter);
    for (const projectHash of sortedProjectStates) {
      hash.update(projectHash);
      hash.update(RushConstants.hashDelimiter);
    }

    const projectStateHash: string = hash.digest('hex');

    return options.buildCacheConfiguration.getCacheEntryId({
      projectName: options.projectConfiguration.project.packageName,
      projectStateHash
    });
  }
}
