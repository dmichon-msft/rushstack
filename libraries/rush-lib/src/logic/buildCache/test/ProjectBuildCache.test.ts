// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { StringBufferTerminalProvider, Terminal } from '@rushstack/node-core-library';
import { BuildCacheConfiguration } from '../../../api/BuildCacheConfiguration';
import { RushProjectConfiguration } from '../../../api/RushProjectConfiguration';
import { IGenerateCacheEntryIdOptions } from '../CacheEntryId';
import { FileSystemBuildCacheProvider } from '../FileSystemBuildCacheProvider';

import { ProjectBuildCache } from '../ProjectBuildCache';

interface ITestOptions {
  enabled: boolean;
  writeAllowed: boolean;
  trackedProjectFiles: string[] | undefined;
}

describe(ProjectBuildCache.name, () => {
  async function prepareSubject(options: Partial<ITestOptions>): Promise<ProjectBuildCache | undefined> {
    const terminal: Terminal = new Terminal(new StringBufferTerminalProvider());
    const hash: string = 'state_hash';

    const subject: ProjectBuildCache | undefined = await ProjectBuildCache.tryGetProjectBuildCache({
      buildCacheConfiguration: {
        buildCacheEnabled: options.hasOwnProperty('enabled') ? options.enabled : true,
        getCacheEntryId: (options: IGenerateCacheEntryIdOptions) =>
          `${options.projectName}/${options.projectStateHash}`,
        localCacheProvider: undefined as unknown as FileSystemBuildCacheProvider,
        cloudCacheProvider: {
          isCacheWriteAllowed: options.hasOwnProperty('writeAllowed') ? options.writeAllowed : false
        }
      } as unknown as BuildCacheConfiguration,
      projectOutputFolderNames: ['dist'],
      projectConfiguration: {
        project: {
          packageName: 'acme-wizard',
          projectRelativeFolder: 'apps/acme-wizard',
          dependencyProjects: []
        }
      } as unknown as RushProjectConfiguration,
      command: 'build',
      trackedProjectFiles: options.hasOwnProperty('trackedProjectFiles') ? options.trackedProjectFiles : [],
      hash,
      terminal,
      phaseName: 'build'
    });

    return subject;
  }

  describe(ProjectBuildCache.tryGetProjectBuildCache.name, () => {
    it('returns a ProjectBuildCache with a calculated cacheId value', async () => {
      const subject: ProjectBuildCache = (await prepareSubject({}))!;
      expect(subject['_cacheId']).toMatchSnapshot();
    });

    it('returns undefined if the tracked file list is undefined', async () => {
      expect(
        await prepareSubject({
          trackedProjectFiles: undefined
        })
      ).toBe(undefined);
    });
  });
});
