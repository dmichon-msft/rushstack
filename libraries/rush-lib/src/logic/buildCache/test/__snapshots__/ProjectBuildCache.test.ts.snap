// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectBuildCache tryGetProjectBuildCache returns a ProjectBuildCache with a calculated cacheId value 1`] = `"acme-wizard/bb7633fef1e8a70f64714aabfeda78b5cd803e3c"`;
